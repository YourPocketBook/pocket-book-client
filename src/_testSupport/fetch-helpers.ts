import fetchMock, {
  MockOptionsMethodDelete,
  MockOptionsMethodGet,
  MockOptionsMethodPost,
  MockOptionsMethodPut,
  MockResponse,
  MockResponseFunction
} from "fetch-mock";

afterEach(fetchMock.reset);

export function authenticatedGet(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodGet
) {
  fetchMock.get(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}

export function authenticatedPut(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPut
) {
  fetchMock.put(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedPost(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPost
) {
  fetchMock.post(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedDelete(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodDelete
) {
  fetchMock.delete(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}
