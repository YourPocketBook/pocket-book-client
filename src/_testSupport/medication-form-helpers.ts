import { fireEvent } from "@testing-library/react";

export const testDefaultState = {
  adviceIfDeclined: "Test Advice 1",
  adviceIfTaken: "Test Advice 2",
  dose: "Test Dose",
  exclusionCriteria: "Test Exclusion",
  form: "Test Form",
  inclusionCriteria: "Test Inclusion",
  indications: "Test Indications",
  name: "Test Name",
  policyDate: "2019-12-10",
  route: "Test Route",
  sideEffects: "Test Side Effects"
};

export function setupState(getByLabelText: (text: string) => HTMLElement) {
  fireEvent.change(getByLabelText("Name"), {
    target: { value: testDefaultState.name }
  });
  fireEvent.change(getByLabelText("Policy Date"), {
    target: { value: testDefaultState.policyDate }
  });
  fireEvent.change(getByLabelText("Indications"), {
    target: { value: testDefaultState.indications }
  });
  fireEvent.change(getByLabelText("Inclusion Criteria"), {
    target: { value: testDefaultState.inclusionCriteria }
  });
  fireEvent.change(getByLabelText("Exclusion Criteria"), {
    target: { value: testDefaultState.exclusionCriteria }
  });
  fireEvent.change(getByLabelText("Name, Forms and Strengths of Medicine"), {
    target: { value: testDefaultState.form }
  });
  fireEvent.change(getByLabelText("Route or Method"), {
    target: { value: testDefaultState.route }
  });
  fireEvent.change(getByLabelText("Dosage"), {
    target: { value: testDefaultState.dose }
  });
  fireEvent.change(getByLabelText("Side Effects"), {
    target: { value: testDefaultState.sideEffects }
  });
  fireEvent.change(getByLabelText("Advice if Taken"), {
    target: { value: testDefaultState.adviceIfTaken }
  });
  fireEvent.change(getByLabelText("Advice if Declined"), {
    target: { value: testDefaultState.adviceIfDeclined }
  });
}

export function checkForClass(
  getByLabelText: (text: string) => HTMLElement,
  cls: "is-valid" | "is-invalid"
) {
  expect(getByLabelText("Name")).toHaveClass(cls);
  expect(getByLabelText("Policy Date")).toHaveClass(cls);
  expect(getByLabelText("Indications")).toHaveClass(cls);
  expect(getByLabelText("Inclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Exclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Name, Forms and Strengths of Medicine")).toHaveClass(
    cls
  );
  expect(getByLabelText("Route or Method")).toHaveClass(cls);
  expect(getByLabelText("Dosage")).toHaveClass(cls);
  expect(getByLabelText("Side Effects")).toHaveClass(cls);
  expect(getByLabelText("Advice if Taken")).toHaveClass(cls);
  expect(getByLabelText("Advice if Declined")).toHaveClass(cls);
}

export function checkState(
  getByLabelText: (text: string) => HTMLElement,
  hasValue: boolean
) {
  expect(getByLabelText("Name")).toHaveValue(
    hasValue ? testDefaultState.name : ""
  );
  expect(getByLabelText("Policy Date")).toHaveValue(
    hasValue ? testDefaultState.policyDate : ""
  );
  expect(getByLabelText("Indications")).toHaveTextContent(
    hasValue ? testDefaultState.indications : ""
  );
  expect(getByLabelText("Inclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.inclusionCriteria : ""
  );
  expect(getByLabelText("Exclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.exclusionCriteria : ""
  );
  expect(
    getByLabelText("Name, Forms and Strengths of Medicine")
  ).toHaveTextContent(hasValue ? testDefaultState.form : "");
  expect(getByLabelText("Route or Method")).toHaveTextContent(
    hasValue ? testDefaultState.route : ""
  );
  expect(getByLabelText("Dosage")).toHaveTextContent(
    hasValue ? testDefaultState.dose : ""
  );
  expect(getByLabelText("Side Effects")).toHaveTextContent(
    hasValue ? testDefaultState.sideEffects : ""
  );
  expect(getByLabelText("Advice if Taken")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfTaken : ""
  );
  expect(getByLabelText("Advice if Declined")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfDeclined : ""
  );
}
