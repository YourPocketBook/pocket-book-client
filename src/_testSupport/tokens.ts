export const inDateNonAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.VmT5YkTG0sUd5sK7skfMdi9kc5XG7Yhx6sqhqMSxhVY";
export const inDateAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInJvbGUiOiJBZG1pbiIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.rSQb2Hv6RfxaXSfO6s3QjLjnYKeqVgps3rPenquh3Nk";
export const expiredToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE0MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInVuaXF1ZV9uYW1lIjoiVGVzdCBVc2VyIiwibmFtZWlkIjoiZmYxZDA4ZTctNDY1Ni00YzI4LTk3NmYtMWMyN2YxOTE0NzdmIiwiZW1haWwiOiJ0ZXN0QHNqYS5vcmcudWsifQ.7bDfHDk9sWXL8boFeZWIdo_GB-4Qj8PvA4HDC49EzNc";
