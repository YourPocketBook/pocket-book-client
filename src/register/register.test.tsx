import "@testing-library/jest-dom/extend-expect";
import { fireEvent, render, wait } from "@testing-library/react";
import deepEqual from "deep-equal";
import fetchMock from "fetch-mock";
import React from "react";
import ReactDOM from "react-dom";
import { Register } from ".";
import "../_testSupport/fetch-helpers";
import { history } from "../history";

jest.mock("zxcvbn");

const endpoint = "/api/register-user";

const testDefaultState = {
  email: "test@sja.org.uk",
  name: "Test Name",
  password: "testPassword",
  updateEmailConsentGiven: true
};

describe("Register Page", () => {
  beforeEach(() => {
    history.push = jest.fn();
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Register />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders correctly", () => {
    const { container } = render(<Register />);

    expect(container).toMatchSnapshot();
  });

  it("defaults to correct values", () => {
    const { getByLabelText } = render(<Register />);

    expect(getByLabelText("Name")).toHaveAttribute("value", "");
    expect(getByLabelText("Email")).toHaveAttribute("value", "");
    expect(getByLabelText("Password")).toHaveAttribute("value", "");
    expect(getByLabelText("Password")).toHaveAttribute("type", "password");
  });

  it("renders valid values correctly", async () => {
    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        deepEqual(JSON.parse((h.body as string) || ""), testDefaultState),
      {
        body: JSON.stringify([{ path: "doses", error: "test" }]),
        status: 400
      }
    );

    const { getByLabelText, getByText } = render(<Register />);

    fireEvent.change(getByLabelText("Name"), {
      target: { value: testDefaultState.name }
    });
    fireEvent.change(getByLabelText("Email"), {
      target: { value: testDefaultState.email }
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: testDefaultState.password }
    });
    fireEvent.click(
      getByLabelText("Privacy Policy", { exact: false, selector: "input" })
    );
    fireEvent.click(
      getByLabelText("Terms", { exact: false, selector: "input" })
    );
    fireEvent.click(
      getByLabelText("consent", { exact: false, selector: "input" })
    );
    fireEvent.click(getByText("Register", { selector: "button" }));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByLabelText("Name")).toHaveClass("is-valid");
    expect(getByLabelText("Email")).toHaveClass("is-valid");
    expect(getByLabelText("Password")).toHaveClass("is-valid");
  });

  it("renders invalid values correctly", () => {
    const { getByLabelText, getByText } = render(<Register />);

    fireEvent.click(getByText("Register", { selector: "button" }));

    expect(getByLabelText("Name")).toHaveClass("is-invalid");
    expect(getByLabelText("Email")).toHaveClass("is-invalid");
    expect(getByLabelText("Password")).toHaveClass("is-invalid");
    expect(
      getByLabelText("Privacy Policy", { exact: false, selector: "input" })
    ).toHaveClass("is-invalid");
    expect(
      getByLabelText("Terms", { exact: false, selector: "input" })
    ).toHaveClass("is-invalid");
  });

  it("renders duplicate email correctly", async () => {
    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        deepEqual(JSON.parse((h.body as string) || ""), testDefaultState),
      {
        body: JSON.stringify([{ path: "email", error: "isInUse" }]),
        status: 400
      }
    );

    const { getByLabelText, getByText } = render(<Register />);

    fireEvent.change(getByLabelText("Name"), {
      target: { value: testDefaultState.name }
    });
    fireEvent.change(getByLabelText("Email"), {
      target: { value: testDefaultState.email }
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: testDefaultState.password }
    });
    fireEvent.click(
      getByLabelText("Privacy Policy", { exact: false, selector: "input" })
    );
    fireEvent.click(
      getByLabelText("Terms", { exact: false, selector: "input" })
    );
    fireEvent.click(
      getByLabelText("consent", { exact: false, selector: "input" })
    );
    fireEvent.click(getByText("Register", { selector: "button" }));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByLabelText("Email")).toHaveClass("is-invalid");
    expect(getByText(/email is already in use/)).toBeDefined();
  });

  it("declares non-SJA email invalid", () => {
    const { getByLabelText, getByText } = render(<Register />);

    fireEvent.change(getByLabelText("Email"), {
      target: { value: "test@gmail.com" }
    });

    fireEvent.click(getByText("Register", { selector: "button" }));

    expect(getByLabelText("Name")).toHaveClass("is-invalid");
    expect(getByLabelText("Email")).toHaveClass("is-invalid");
    expect(getByLabelText("Password")).toHaveClass("is-invalid");
    expect(
      getByLabelText("Privacy Policy", { exact: false, selector: "input" })
    ).toHaveClass("is-invalid");
    expect(
      getByLabelText("Terms", { exact: false, selector: "input" })
    ).toHaveClass("is-invalid");

    expect(getByLabelText("Email")).toHaveClass("is-invalid");
    expect(getByText(/give your SJA email/)).toBeDefined();
  });

  it("renders server error correctly", async () => {
    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        deepEqual(JSON.parse((h.body as string) || ""), testDefaultState),
      {
        status: 500
      }
    );

    const { getByLabelText, getByText } = render(<Register />);

    fireEvent.change(getByLabelText("Name"), {
      target: { value: testDefaultState.name }
    });
    fireEvent.change(getByLabelText("Email"), {
      target: { value: testDefaultState.email }
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: testDefaultState.password }
    });
    fireEvent.click(
      getByLabelText("Privacy Policy", { exact: false, selector: "input" })
    );
    fireEvent.click(
      getByLabelText("Terms", { exact: false, selector: "input" })
    );
    fireEvent.click(
      getByLabelText("consent", { exact: false, selector: "input" })
    );
    fireEvent.click(getByText("Register", { selector: "button" }));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText(/The server has had an issue/)).toBeInTheDocument();
  });

  it("renders in progress correctly", async () => {
    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        deepEqual(JSON.parse((h.body as string) || ""), testDefaultState),
      // tslint:disable-next-line: no-empty
      new Promise(() => {})
    );

    const { getByLabelText, getByText } = render(<Register />);

    fireEvent.change(getByLabelText("Name"), {
      target: { value: testDefaultState.name }
    });
    fireEvent.change(getByLabelText("Email"), {
      target: { value: testDefaultState.email }
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: testDefaultState.password }
    });
    fireEvent.click(
      getByLabelText("Privacy Policy", { exact: false, selector: "input" })
    );
    fireEvent.click(
      getByLabelText("Terms", { exact: false, selector: "input" })
    );
    fireEvent.click(
      getByLabelText("consent", { exact: false, selector: "input" })
    );
    fireEvent.click(getByText("Register", { selector: "button" }));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText("Register", { selector: "button" })).toBeDisabled();
  });

  it("toggles show password when button is clicked", () => {
    const { getByText, getByLabelText } = render(<Register />);

    fireEvent.click(getByText("Show Password"));
    expect(getByLabelText("Password")).toHaveAttribute("type", "text");

    fireEvent.click(getByText("Hide Password"));
    expect(getByLabelText("Password")).toHaveAttribute("type", "password");
  });

  it("redirects on successful form submit", async () => {
    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        deepEqual(JSON.parse((h.body as string) || ""), testDefaultState),
      {
        status: 204
      }
    );

    const { getByLabelText, getByText } = render(<Register />);

    fireEvent.change(getByLabelText("Name"), {
      target: { value: testDefaultState.name }
    });
    fireEvent.change(getByLabelText("Email"), {
      target: { value: testDefaultState.email }
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: testDefaultState.password }
    });
    fireEvent.click(
      getByLabelText("Privacy Policy", { exact: false, selector: "input" })
    );
    fireEvent.click(
      getByLabelText("Terms", { exact: false, selector: "input" })
    );
    fireEvent.click(
      getByLabelText("consent", { exact: false, selector: "input" })
    );
    fireEvent.click(getByText("Register", { selector: "button" }));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(history.push).toBeCalledWith("/register-success");
  });
});
