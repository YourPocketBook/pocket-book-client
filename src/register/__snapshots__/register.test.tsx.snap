// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Page renders correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
    </nav>
    <div
      class="centreContainer container"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-6 offset-lg-3"
        >
          <div
            class="centreBox"
          >
            <div
              class="container"
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  <h1
                    class="text-center"
                  >
                    Register
                  </h1>
                </div>
              </div>
              <form
                class=""
                id="register-form"
                novalidate=""
              >
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <div
                      class="form-group"
                    >
                      <label
                        class=""
                        for="name"
                      >
                        Name
                      </label>
                      <input
                        autocomplete="name"
                        class="form-control"
                        id="name"
                        name="name"
                        placeholder="Name"
                        type="text"
                        value=""
                      />
                      <div
                        class="invalid-feedback"
                      >
                        You need to give your name.
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <div
                      class="form-group"
                    >
                      <label
                        class=""
                        for="email"
                      >
                        Email
                      </label>
                      <input
                        autocomplete="email"
                        class="form-control"
                        id="email"
                        name="email"
                        placeholder="Email"
                        type="email"
                        value=""
                      />
                      <div
                        class="invalid-feedback"
                      >
                        You need to give your SJA email
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <div
                      class="form-group"
                    >
                      <label
                        class=""
                        for="password"
                      >
                        Password
                      </label>
                      <input
                        autocomplete="new-password"
                        class="form-control"
                        id="password"
                        name="password"
                        placeholder="Password"
                        type="password"
                        value=""
                      />
                      <div
                        class="invalid-feedback"
                      >
                        You need to enter a password.
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="row"
                >
                  <div
                    class="text-right col-md-9 offset-md-3"
                  >
                    <button
                      class="text-muted btn btn-link"
                      id="showPassword"
                      type="button"
                    >
                      Show Password
                    </button>
                  </div>
                </div>
                <div
                  class="form-check"
                >
                  <label
                    class="form-check-label"
                  >
                    <input
                      class="form-check-input"
                      id="acceptPrivacy"
                      type="checkbox"
                    />
                    I have read and accept the
                     
                    <a
                      class=""
                      href="/privacy"
                    >
                      Privacy Policy
                    </a>
                    .
                    <div
                      class="invalid-feedback"
                    >
                      You can't register if you don't accept this policy.
                    </div>
                  </label>
                </div>
                <div
                  class="form-check"
                >
                  <label
                    class="form-check-label"
                  >
                    <input
                      class="form-check-input"
                      id="acceptTerms"
                      type="checkbox"
                    />
                    I have read and accept the
                     
                    <a
                      class=""
                      href="/privacy"
                    >
                      Terms and Conditions of Use
                    </a>
                    .
                    <div
                      class="invalid-feedback"
                    >
                      You can't register if you don't accept this policy.
                    </div>
                  </label>
                </div>
                <div
                  class="form-check"
                >
                  <label
                    class="text-small form-check-label"
                  >
                    <input
                      class="form-check-input"
                      id="acceptEmails"
                      type="checkbox"
                    />
                    I consent to receive occasional emails when this site is updated. This is optional.
                  </label>
                </div>
                <button
                  class="btn btn-dark-green btn-block"
                  type="submit"
                >
                  Register
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright Â© 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
